theano/gof/wrapper.py

```diff
@@ -0,0 +1,439 @@
+            void extract_%(attribute_name)s(PyObject* py_%(attribute_name)s) {
+                %(extract_code)s
+            }
+            """ % {
+                'attribute_name': attribute_name,
+                'extract_code': type_instance.c_extract(attribute_name, sub)
+            })
+
+        struct_declare = '\n'.join(c_declare_list)
+        struct_init = '\n'.join(c_init_list)
+        struct_cleanup = '\n'.join(c_cleanup_list)
+        struct_extract = '\n\n'.join(c_extract_list)
+        struct_extract_method = """
+        void extract(PyObject* object, int field_pos) {
+            switch(field_pos) {
+                // Extraction cases.
+                %s
+                // Default case.
+                default:
+                    PyErr_Format(PyExc_TypeError, "Wrapper: no extraction defined for a field %%d.", field_pos);
+                    this->setErrorOccurred();
+                    break;
+            }
+        }
+        """ % ('\n'.join(
+            [('case %d: extract_%s(object); break;' % (i, self.fields[i])) for i in range(self.length)])
+        )
+        return """
+        #ifndef %(struct_name_defined)s
+        #define %(struct_name_defined)s
+        struct %(struct_name)s {
+            /* Attributes, */
+            int %(struct_name)s_error;
+            %(struct_declare)s
+
+            /* Constructor. */
+            %(struct_name)s() {
+                %(struct_name)s_error = 0;
+                %(struct_init)s
+            }
+
+            /* Destructor. */
+            ~%(struct_name)s() {
+                // cleanup() is defined below.
+                cleanup();
+            }
+
+            /* Cleanup method. */
+            void cleanup() {
+                %(struct_cleanup)s
+            }
+
+            /* Extraction methods. */
+            %(struct_extract)s
+
+            /* Extract method. */
+            %(struct_extract_method)s
+
+            /* Other methods. */
+            void setErrorOccurred() {
+                ++%(struct_name)s_error;
+            }
+            int errorOccurred() {
+                return %(struct_name)s_error;
+            }
+        };
+        #endif
+        """ % locals()
+
+    def c_code_cache_version(self):
+        wrapper_c_code_version = (1, 6)
+        return (wrapper_c_code_version, tuple(t.c_code_cache_version() for t in self.types))
+
+    # As this struct has constructor and destructor, it could be instanciated on stack,
+    # but current implementations of C ops will then pass the instance by value at functions,
+    # so it's better to work directly with pointers.
+
+    def c_declare(self, name, sub, check_input=True):
+        struct_name = self.name
+        return """
+        %(struct_name)s* %(name)s;
+        """ % locals()
+
+    def c_init(self, name, sub):
+        # NB: It seems c_init() is not called for an op param.
+        # So the real initialization is done at top of c_extract.
+        return """
```

Comment:
Yes, that was my first approach (declare struct on stack) ! But I changed it to pointers because of 2 considerations:
1) If struct is an object rather than a pointer, then it is copied when passed to APPLY_SPECIFIC main function for COps. We should then expect that wrapper objects are not too big so that they are copied faster enough. But I think it is not so problematic, finally.
2) The original wrapper and the copy passed to APPLY_SPECIFIC main function in COps will share pointers values (like PyObject* pointers). Then, at the end of APPLY_SPECIFIC function, the destructor will be called on the wrapper copy, that will call the cleanup function, that will clean the PyObjects and other memory data. Then, we must be sure that the original wrapper will not be reused after the main function is called, as its pointers should have been cleaned (e.g. PyObject pointers into wrapper would have been decremented and set to NULL, etc.). For this point, are we sure ? If yes, I will go back to structs on stack (which is indeed better, in my opinion !).